src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,747 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+using static System.Array;
+using static System.Char;
+using static System.Int32;
+
+#nullable enable
+
+namespace System.Reflection.Metadata
+{
+    internal static class TypeNameParserHelpers
+    {
+        internal const int SZArray = -1;
+        internal const int Pointer = -2;
+        internal const int ByRef = -3;
+        private const char EscapeCharacter = '\\';
+#if NET8_0_OR_GREATER
+        private static readonly SearchValues<char> _endOfTypeNameDelimitersSearchValues = SearchValues.Create(".+");
+        private static readonly SearchValues<char> _endOfFullTypeNameDelimitersSearchValues = SearchValues.Create("[]&*,+\\");
+#endif
+
+        /// <returns>
+        /// <para>Negative value for invalid type names.</para>
+        /// <para>Zero for valid non-generic type names.</para>
+        /// <para>Positive value for valid generic type names.</para>
+        /// </returns>
+        internal static int GetGenericArgumentCount(ReadOnlySpan<char> fullTypeName)
+        {
+            const int ShortestInvalidTypeName = 2; // Back tick and one digit. Example: "`1"
+            if (fullTypeName.Length < ShortestInvalidTypeName || !IsAsciiDigit(fullTypeName[fullTypeName.Length - 1]))
+            {
+                return 0;
+            }
+
+            int backtickIndex = fullTypeName.Length - 2; // we already know it's true for the last one
+            for (; backtickIndex >= 0; backtickIndex--)
+            {
+                if (fullTypeName[backtickIndex] == '`')
+                {
+                    if (backtickIndex == 0)
+                    {
+                        return -1; // illegal name, example "`1"
+                    }
+                    else if (fullTypeName[backtickIndex - 1] == EscapeCharacter)
+                    {
+                        return 0; // legal name, but not a generic type definition. Example: "Escaped\\`1"
+                    }
+                    else if (TryParse(fullTypeName.Slice(backtickIndex + 1), out int value))
+                    {
+                        // From C# 2.0 language spec: 8.16.3 Multiple type parameters Generic type declarations can have any number of type parameters.
+                        if (value > MaxLength)
+                        {
+                            // But.. it's impossible to create a type with more than Array.MaxLength.
+                            // OOM is also not welcomed in the parser!
```

Comment:
If the parser allocates anything, it could possibly OOM.

Reply:
I wanted to avoid a situation, where we preallocate arrays of generic arguments for ``TypeName`$MoreThanArrayMaxLength`` where we know it's guaranteed to get OOM.

Reply:
I agree with Stephen that there are many places in the parser that can OOM for extremely long names and it is not clear why we are special-casing this one. This code looks like unnecessary overengineering.
 
For the hardened cases, the potential OOM for extremely long names should be covered by the MaxNode limit.

Reply:
I went too far with my "security paranoia", I am going to remove that. Thanks!